apm-agent-plugins/apm-log-shader-plugin/apm-log-shader-plugin-common/src/test/java/co/elastic/apm/agent/log/shader/LogShadingInstrumentationTest.java

```diff
@@ -0,0 +1,241 @@
+        ArrayList<JsonNode> ecsLogLines = new ArrayList<>();
+        try (Stream<String> stream = Files.lines(Paths.get(shadeLogFilePath))) {
+            stream.forEach(line -> {
+                try {
+                    ecsLogLines.add(objectMapper.readTree(line));
+                } catch (JsonProcessingException e) {
+                    e.printStackTrace();
+                }
+            });
+        }
+        return ecsLogLines;
+    }
+
+    @Nonnull
+    private ArrayList<String[]> readRawLogLines() throws IOException {
+        ArrayList<String[]> rawLogLines;
+        try (Stream<String> stream = Files.lines(getOriginalLogFilePath())) {
+            rawLogLines = stream.map(line -> line.split("\\s+")).collect(Collectors.toCollection(ArrayList::new));
+        }
+        return rawLogLines;
+    }
+
+    @Nonnull
+    private Path getOriginalLogFilePath() {
+        return Paths.get(logger.getLogFilePath());
+    }
+
+    @Nonnull
+    private String getShadeLogFilePath() {
+        return Utils.computeShadeLogFilePath(logger.getLogFilePath());
+    }
+
+    private void verifyEcsFormat(String[] splitRawLogLine, JsonNode ecsLogLineTree, @Nullable String traceId) throws ParseException {
+        SimpleDateFormat timestampFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+        Date rawTimestamp = timestampFormat.parse(splitRawLogLine[0]);
+        timestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+        Date ecsTimestamp = timestampFormat.parse(ecsLogLineTree.get("@timestamp").textValue());
+        assertThat(rawTimestamp).isEqualTo(ecsTimestamp);
+        assertThat(splitRawLogLine[1]).isEqualTo(ecsLogLineTree.get("process.thread.name").textValue());
+        assertThat(splitRawLogLine[2]).isEqualTo(ecsLogLineTree.get("log.level").textValue());
+        assertThat(splitRawLogLine[3]).isEqualTo(ecsLogLineTree.get("log.logger").textValue());
+        assertThat(splitRawLogLine[4]).isEqualTo(ecsLogLineTree.get("message").textValue());
+        assertThat(ecsLogLineTree.get("service.name").textValue()).isEqualTo(tracer.getMetaData().getService().getName());
+        if (traceId != null) {
+            assertThat(ecsLogLineTree.get("trace.id").textValue()).isEqualTo(traceId);
+        } else {
+            assertThat(ecsLogLineTree.get("trace.id")).isNull();
+        }
+    }
+
+    /**
+     * Disabled by default, as this is a very slow test. Can be used for manual testing of shade file rolling.
+     * Note: logback and log4j2 rollover before appending an event, which means the two log files will contain messages.
+     * As opposed to those, log4j1 rolls over after appending an event, which means that the active log file (log4j1.log)
+     * will be empty when the test ends.
+     */
+     //@Test
+     public void testShadeLogRolling() throws IOException {
+         when(config.getConfig(LoggingConfiguration.class).getLogFileSize()).thenReturn(100L);
+         logger.trace("First line");
+         sleep();
```

Comment:
An idea to make the test faster
```suggestion
         await().untilAsserted(() -> new File(getShadeLogFilePath()).length() == 0);
```

Reply:
Unfortunately, this is only possible for log4j1, which rolls the file at the end of the event logging. 
It is also possible in log4j2 if I set `immediateFlush=false`, but this fails all other tests. I used two different loggers, one for the rolling test (with `immediateFlush==false`) and one for the rest, but it ended up being very close in duration to using a sleep of 50ms, which seems safe in log4j2. 
Logback required a very long sleep, but it works well and fast with `immediateFlush==false`.

Eventually I got to a state where these tests take 200-600ms, which is good enough to enable by default. If they will prove to be flaky we can extend the sleep a bit or disable.

Reply:
Already proved flaky in the first test 🙁 
I raised sleep time to 4X. Let's see where it takes us.

Reply:
This test was very useful for me to test our log file rolling configurations. It will be as useful in the future if we would like to change something like our rolling-decision strategy. However, it is not as important for continuous-integration testing, which is why I disabled it and still left it. 
It now adds about 1.5 seconds to the build for the three logging frameworks combined, so as long as it is not flaky, we can leave it on and disable later if required.